src/Data/LLVM/BitCode/GetBits.hs

```diff
@@ -125,112 +117,233 @@ instance MonadPlus GetBits where
 -- with the bit position following the extraction.
 
-extractFromByteString :: NumBits {-^ the last bit accessible in the ByteString -}
-                      -> NumBits {-^ the bit to start extraction at -}
-                      -> NumBits {-^ the number of bits to extract -}
+-- There are two implementations: one builds an integer value by shifting bits,
+-- then shifts and masks the result to get the final value.  The other uses
+-- unlifted values and avoids the final shift by being smarter about individual
+-- compositions.  Their functionality should be identical, but it may be easier
+-- to debug the first.
+
+-- extractFromByteString bitLimit startBit numBits bs =
+--   let Bytes' s8 = fst (bitsToBytes startBit)
+--       Bytes' r8 = fst (bitsToBytes numBits)
+--       rcnt = r8 + 2 -- 2 == pre-shift overflow byte on either side
+
+--       -- Extract the relevant bits from the ByteCode, with padding to byte
+--       -- boundaries into ws.
+--       ws = BS.take rcnt $ BS.drop s8 bs
+
+--       -- Combine the extracted bytes into an Integer value in wi.
+--       wi = BS.foldr (\w a -> a `shiftL` 8 .|. fromIntegral w) (0::Int) ws
+
+--       -- Mask is 0-bit based set of bits wanted in the result
+--       mask = ((1::Int) `shiftL` bitCount numBits) - 1
+
+--       -- Shift the desired value down to byte alignment and then discard any
+--       -- excess high bits.
+--       vi = wi `shiftR` (bitCount startBit .&. 7) .&. mask
+
+--       updPos = addBitCounts startBit numBits
+--   in if updPos > bitLimit
+--      then Left ("Attempt to read bits past limit (newPos="
+--                 <> show updPos <> ", limit=" <> show bitLimit <> ")"
+--                )
+--      else Right (vi, updPos)
+
+extractFromByteString :: Int# {-^ the last bit accessible in the ByteString -}
+                      -> Int# {-^ the bit to start extraction at -}
+                      -> Int# {-^ the number of bits to extract -}
                       -> ByteString {-^ the ByteString to extract from -}
-                      -> Either String (Integer, NumBits)
-extractFromByteString bitLimit startBit numBits bs =
-  let Bytes' s8 = fst (bitsToBytes startBit)
-      Bytes' r8 = fst (bitsToBytes numBits)
-      rcnt = r8 + 2 -- 2 == pre-shift overflow byte on either side
-
-      -- Extract the relevant bits from the ByteCode, with padding to byte
-      -- boundaries into ws.
-      ws = BS.take rcnt $ BS.drop s8 bs
-
-      -- Combine the extracted bytes into an Integer value in wi.
-      wi = BS.foldr (\w a -> a `shiftL` 8 .|. fromIntegral w) (0::Integer) ws
-
-      -- Mask is 0-bit based set of bits wanted in the result
-      mask = ((1::Integer) `shiftL` bitCount numBits) - 1
-
-      -- Shift the desired value down to byte alignment and then discard any
-      -- excess high bits.
-      vi = wi `shiftR` (bitCount startBit .&. 7) .&. mask
+                      -> Either String (() -> (# Int#, Int# #))
```

Comment:
The use of `() -> (# Int#, Int# #)` here is rather awkward. I get the sense that the `()` argument was only added to make it work with `Either`, which expects boxed arguments. Why not use `UnboxedSums` instead? Something like this, to be more precise:

```hs
(# String | (# Int#, Int# #) #)
```

Reply:
That is indeed the case.  I'll re-address this in a future PR.